File: Python/Problem 4.py
class Solution(object):
    def isMatch(self, s, p):
        i_s = 0
        i_p = 0
        res = false

        for i in range(0, len(s)):
            if 0 == len(p):
                return 0 == len(s)

            res = i_s < len(s) and (p[i_p] == s[i_s] or p[i_p] == '.')

            if !res:
                break;

            if i_p+1 < len(p) and p[i_p+1] == '*':
                if i_s+1 < len(s) and s[i_s+1] != p[i_p]:
                    i_p += 2
                else:
                    i_s++
            else:
                i_p++
                i_s++
        return res

File: Java/Problem 4.java
class Solution {
    public boolean isMatch(String s, String p) {
        if (0 == p.length()){
            return 0 == s.length();
        }
    
        int i_s = 0;
        int i_p = 0;
        Boolean res = false;

        while (i_p < p.length()){
        
            res = (i_s < length(s) && (p.charAt(i_p) == s.charAt(i_s) or p.charAt(i_p) == '.'));

            if (!res){
                return res;
            }

            if (i_p+1 < p.length() && p.charAt(i_p+1) == '*'){
                if (i_s+1 < s.length() && s.charAt(i_s+1) != p.charAt(i_p)){
                    i_p += 2;
                }else{
                    i_s++;
                }
            }else{
                i_p++;
                i_s++;
            }
        }
        return i_s == s.length();
    }
}
